torch/csrc/api/src/optim/lbfgs.cpp

```diff
@@ -20,6 +20,10 @@ LBFGSOptions::LBFGSOptions(double learning_rate)
 Tensor LBFGS::gather_flat_grad() {
   std::vector<Tensor> views;
   for (auto& parameter : parameters_) {
+    if (!parameter.grad().defined()) {
+      continue;
+    }
```

Comment:
It seems that the Python implementation for this part is https://github.com/pytorch/pytorch/blob/7c472ec5974cc638cb697b4523d88324aee38fa9/torch/optim/lbfgs.py#L250-L251
Do you think we can do the same for C++ implementation?

Reply:
Thank you for your review! I have modified the code in `torch/csrc/api/src/optim/lbfgs.cpp` according to the code in python `pytorch/torch/optim/lbfgs.py`:
```
...
    if (!parameter.grad().defined()) {
      views.push_back(parameter.new_empty({parameter.numel()}).zero_());
    }
    else if (parameter.grad().is_sparse()) {
      views.push_back(parameter.grad().to_dense().view(-1));
    }
    else {
      views.push_back(parameter.grad().view(-1));
    }
...
```
Now the test in #27605 can run without error on both CPU and GPU:
```
#Output
Epoch: 0 Batch: 0 loss: 0.903995
[ Variable[CUDAFloatType]{} ]
Epoch: 0 Batch: 0 loss: 1.413
[ Variable[CUDAFloatType]{} ]
Epoch: 1 Batch: 0 loss: 0.731126
[ Variable[CUDAFloatType]{} ]
Epoch: 1 Batch: 0 loss: 0.771402
[ Variable[CUDAFloatType]{} ]
Epoch: 2 Batch: 0 loss: 0.482919
[ Variable[CUDAFloatType]{} ]
Epoch: 2 Batch: 0 loss: 0.296223
[ Variable[CUDAFloatType]{} ]
Epoch: 3 Batch: 0 loss: 0.22991
[ Variable[CUDAFloatType]{} ]
Epoch: 3 Batch: 0 loss: 0.0694233
[ Variable[CUDAFloatType]{} ]
Epoch: 4 Batch: 0 loss: 0.0918769
...
```

However, another error occured (caused by `torch::optim::LBFGS::add_grad`) if there is a sparse tensor in the model:
```
terminate called after throwing an instance of 'c10::Error'
  what():  add(sparse, dense) is not supported. Use add(dense, sparse) instead. (add_out_sparse_cuda at /home/aurora/Softwares/pytorch/aten/src/ATen/native/sparse/cuda/SparseCUDATensorMath.cu:390)
```
I tested using python, and got a similar same error:
```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python3.5/dist-packages/torch/optim/lbfgs.py", line 427, in step
    self._add_grad(t, d)
  File "/usr/local/lib/python3.5/dist-packages/torch/optim/lbfgs.py", line 264, in _add_grad
    p.data.add_(step_size, update[offset:offset + numel].view_as(p.data))
RuntimeError: add(sparse, dense) is not supported. Use add(dense, sparse) instead.
```
The sparse tensor is added to the model by adding a `sparse_fc3` tensor:
```
sparse_fc3 = register_parameter("sparse_fc3", (3 * torch::rand({128,1})).to(torch::kInt).to(torch::kFloat).to_sparse(), true);
```
and by adding
```
x = torch::mm(x, sparse_fc3.to_dense());
```
in the `forward()` function of the model.
I will try to work on this problem.

Reply:
Thanks so much for the investigation! I think sparse tensor support is still experimental within PyTorch, and we will need to also fix the Python version of `LBFGS` to make the sparse use case work. I think this PR as it stands now fixes the problem mentioned in the original issue, and we can merge it now and fix the sparse tensor issue (if desired) in another PR.